[BEAM-801, BEAM-787] Improvements to Splittable DoFn

jkff opened the pull request:
Makes Splittable DoFn be more like a real DoFn:
- Adds support for side inputs and outputs to SDF
- Teaches `ProcessFn` to work with exploded windows inside the
  `KeyedWorkItem`. It works with them by un-exploding the windows
  in the `Iterable<WindowedValue<ElementAndRestriction>>` into a
  single `WindowedValue`, since the values and timestamps are
  guaranteed to be the same.

Makes SplittableParDo.ProcessFn not use the (now unavailable)
OldDoFn state and timers API:
- Makes `ProcessFn` be a primitive transform with its own
  `ParDoEvaluator`. As a nice side effect, this enables the runner to
  provide additional hooks into it - e.g. for giving the runner access
  to the restriction tracker (in later PRs)
- For consistency, moves declaration of `GBKIntoKeyedWorkItems`
  primitive transform into `SplittableParDo`, alongside the
  `SplittableProcessElements` transform
- Preserves compressed representation of `WindowedValue`'s in
  `PushbackSideInputDoFnRunner`

TODO: Tests for new features
TODO: Trigger on the GBK between initial split and ProcessFn
should be like the ReshuffleTrigger.
R: @kennknowles @tgroh

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunner.java:
```diff
@@ -74,17 +74,29 @@ public void startBundle() {
+    if (!readyWindows.isEmpty()) {
+      processElement(
+          WindowedValue.of(
+              elem.getValue(), elem.getTimestamp(), readyWindows, elem.getPane()));
```
I like this. Can you update the javadoc on `SimpleDoFnRunner` to make it clear that it owns exploding things when the value observes the window? I've felt that it was somewhat unclear before now.

jkff replied:
Done.

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -78,27 +90,92 @@
-  private final DoFnSignature signature;
+public class SplittableParDo {
+  /** The version of {@link SplittableParDo} for a {@link ParDo} with a single main output. */
+  public static class Bound<InputT, OutputT, RestrictionT>
```
[BEAM-447](https://issues.apache.org/jira/browse/BEAM-447) "stop referring to types as `Bound`/`Unbound`".

Especially here where there is no such duality.

jkff replied:
Done.

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -225,49 +376,73 @@ public void processElement(ProcessContext context) {
     }
 
-    @Override
+    public void setStateInternalsFactory(StateInternalsFactory<String> stateInternalsFactory) {
```
Should be able to pass these to the constructor. That's actually almost the whole reason `StateInternalsFactory` is a factory.

jkff replied:
See comment below.

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -225,49 +376,73 @@ public void processElement(ProcessContext context) {
+      this.timerInternalsFactory = timerInternalsFactory;
+    }
+
+    public void setOutputManager(DoFnRunners.OutputManager outputManager) {
```
A `DoFn` should not depend on `DoFnRunner` and definitely shouldn't have a handle on `DoFnRunners.OutputManager`. A `DoFnRunner` is just one way of running a `DoFn` that we provide as a convenience.

jkff replied:
Done with the help of #1353 .

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -225,49 +376,73 @@ public void processElement(ProcessContext context) {
         // The element and restriction are available in c.element().
         WindowedValue<ElementAndRestriction<InputT, RestrictionT>> windowedValue =
-            Iterables.getOnlyElement(c.element().elementsIterable());
+            implodeWindows(c.element().elementsIterable());
```
In addition to explain that you are allowed to do this, please explain why you do it (to protect against future edits).

jkff replied:
Done.

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -339,17 +540,17 @@ public PaneInfo pane() {
-              .windowingInternals()
-              .outputWindowedValue(
-                  output, element.getTimestamp(), element.getWindows(), element.getPane());
+          outputManager.output(
```
To remove the dependence on `OutputManager`, here are ideas:

 - Can you make the output type simply be `WindowedValue` (with appropriate compensation afterwards). That would be an accurate description of the `DoFn` anyhow.
 - If you need to output to a particular window, `ReduceFnRunner` also needs a one-method interface with `outputWindowedValue(...)` to remove its dependence on `WindowingInternals` so you could create and use it here.
 - If you need to output a compressed representation, then you'll need a new interface different from the above, as `ReduceFnRunner` only outputs to one window at a time.

jkff replied:
Done in #1353.

kennknowles commented on runners/core-java/src/main/java/org/apache/beam/runners/core/SplittableParDo.java:
```diff
@@ -369,17 +570,17 @@ public PipelineOptions getPipelineOptions() {
-          // "internals.sideOutputWindowedValue".
-          throw new UnsupportedOperationException(
-              "Side outputs not yet supported by splittable DoFn");
+          outputManager.output(
```
There should be a `sideOutputWindowedValue` to match `outputWindowedValue`, for sure, on whatever we end up with.

kennknowles commented on runners/core-java/src/test/java/org/apache/beam/runners/core/PushbackSideInputDoFnRunnerTest.java:
```diff
@@ -162,9 +161,14 @@ public void processElementSideInputNotReadySomeWindows() {
+            WindowedValue.of(
+                2,
+                new Instant(-2),
+                ImmutableList.of(littleWindow, bigWindow),
```
I am OK with, but not excited about, the idea of making maintenance of the compressed representation part of the spec. In general I'd like to be as liberal as possible about allowing different areas of the SDK and runners-core choose between alternative equivalent representations.

kennknowles commented on runners/core-java/src/test/java/org/apache/beam/runners/core/SplittableParDoTest.java:
```diff
@@ -146,18 +151,23 @@ public void testBoundednessForUnboundedFn() {
         PCollection.IsBounded.UNBOUNDED,
         makeUnboundedCollection(pipeline)
-            .apply("unbounded to unbounded", new SplittableParDo<>(unboundedFn))
+            .apply("unbounded to unbounded", new SplittableParDo.Bound<>(ParDo.of(unboundedFn)))
```
Referring to my prior comments, all the things `set` just below should be passed to this constructor.

jkff replied:
Unfortunately that's not possible - there's a cyclic dependency between DoFnTester and ProcessFn: DoFnTester is constructed with ProcessFn, but ProcessFn needs an OutputWindowedValue that will refer to DoFnTester.

Well, it's possible, with some trickery, but it seems not worth it. I think setters are a fine way of wiring together objects with a cyclic dependency.

kennknowles replied:
This is quite an interesting situation. `DoFnTester` is for users, and it plays the role of the runner/worker in this regard.

So for a user who actually wants to test their splittable `DoFn`, the conversion to use any of the technology you have included should occur within `DoFnTester`, privately. But, of course, `DoFnTester` cannot access `runners-core`. I am not familiar enough with the preceding work on splittable `DoFn` to know how this should be resolved.

Often (as you can see from the internals in `DoFnTester`) we end up needing a slight variant on the details of what would otherwise come from `runners-core`. So in this case having a near-duplicate implementation there is not so bad. In the case of a splittable `DoFn` I would expect that the `DoFnTester` code to exercise splittability would diverge even further from this particular implementation in terms of `StateInternals`, etc.

If we stick to hacking on `setXYZ` methods, I would prefer to add `DoFnTester.getOutputWindowedValue(...)` accessor and `ProcessFn.setOutputWindowedValue(...)`, or appropriate variations. This will limit the use of this pattern to just testing, while all worker-side uses will be able to pass things through the constructor without a mutator.

jkff replied:
Note that in this test, I'm testing ProcessFn, rather than a user's splittable DoFn. Agreed that I'll need to make DoFnTester usable with user splittable DoFn's at some point.

Unfortunately I can't have DoFnTester.getOutputWindowedValue, because OutputWindowedValue is declared in runners-core and I'd like to keep it there. It is not a useful part of the SDK; like ProcessFn, it should be part of runners-core.

kennknowles commented on sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/TimerInternalsFactory.java:
```diff
@@ -0,0 +1,36 @@
+ * albeit at execution time, it is marked {@link Serializable}.
+ */
+@Experimental(Kind.STATE)
+public interface TimerInternalsFactory<K> {
```
+1 we need this

jkff commented:
This depends on #1353 - look at the two other commits.
Also, the original title of the PR is not quite as entertaining anymore :(
Still working on tests.

kennknowles commented:
I've left some comments on particular commits, but between conflicts and the underlying PR, just to clear up my tooling, I am going to defer reviewer assignment for now.

R: -@kennknowles

jkff commented:
Fixed the various build warnings and a test error, PTAL.

jkff commented:
The remaining test error is fixed by https://github.com/apache/incubator-beam/pull/1367

jkff commented:
Rebased onto recent changes merging single- and multi-output ParDo evaluator factories. SplittableParDo now overrides only the multi-output ParDo (because single-output is overridden into multi-output).

However, I had to restore TransformHooks in ParDoEvaluatorFactory, because splittable and non-splittable are treated differently. But at least I could remove one type argument (POutput type - which is now always PCollectionTuple).

jkff commented:
...Aaaand while addressing your comment, I realized that things are a little bit more complex, and SDF is failing to invoke the actual fn's Setup/Teardown/StartBundle/FinishBundle methods. Oh well. I'll work on this after Thanksgiving.

jkff commented:
OK, should be all good. PTAL!

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java:
```diff
@@ -0,0 +1,191 @@
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SplittableProcessElementsEvaluatorFactory.class);
+
+  private final LoadingCache<DoFn<?, ?>, DoFnLifecycleManager> fnClones;
```
Can you really not re-use the logic in `ParDoEvaluatorFactory` rather than copy and paste?

jkff replied:
Done; moved a few things around and now the whole ProcessFn is cacheable.

kennknowles commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java:
```diff
@@ -0,0 +1,191 @@
+                  transform.getSideOutputTags().getAll(),
+                  application.getOutput().getAll());
+
+      processFn.setStateInternalsFactory(
```
I understand that you need to have `setStateInternalsFactory` in order to use this with `DoFnTester`. But all non-test code should follow the normal pattern of passing dependencies to the constructor.

jkff replied:
If I'm using ParDoEvaluator to evaluate processFn at all, I'm not sure how I can pass stateInternalsFactory, timerInternalsFactory and outputWindowedValue to the constructor - they are different in every bundle. Do you have suggestions?

kennknowles replied:
The evaluator, such as `ParDoEvaluator`, is one per bundle. So `SplittableProcessElementsEvaluatorFactory.createEvaluator` is safe to pass them right here.

kennknowles commented on runners/direct-java/src/test/java/org/apache/beam/runners/direct/SplittableDoFnTest.java:
```diff
@@ -228,4 +236,172 @@ public void testPairWithIndexWindowedTimestamped() throws ClassNotFoundException
+    }
+  }
+
+  @Test
```
This should be `@NeedsRunner`

jkff replied:
This whole test class should be, but I'd rather do that in a separate PR (move it to runners-core and exclude it from all runners but the direct one, like you did with Stateful ParDo tests).

kennknowles commented:
R: @kennknowles (since I removed before)

jkff commented:
@kennknowles says he'd like to pass the review to @tgroh (with some followup offline comments), so
R: @tgroh

tgroh commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/SplittableProcessElementsEvaluatorFactory.java:
```diff
@@ -0,0 +1,143 @@
+                stepContext,
+                fnManager);
+
+    processFn.setStateInternalsFactory(
```
The `ProcessFn` is to all appearances a System DoFn, as it requires intervention from the system to be properly executed. As such, you should construct it on demand, as it's an implementation detail of `SplittableDoFn` - which means you can construct a new instance within the evaluator for each bundle. The `SplittableParDo` should instead carry as its payload the `SplittableDoFn`, and the `ProcessFn` can be constructed with appropriate `StateInternals`, `TimerInternals`, and `OutputWindowedValue` (`WindowedOutputReciever`?). This is almost the same pattern we formerly used for `GroupAlsoByWindowsDoFn` before that code was entirely migrated to be entirely a primitive in https://github.com/apache/incubator-beam/pull/1251

You can still invoke the methods via `ParDoEvaluator` (though I think generally that's probably discouraged, as the runner will probably want to interact with a `SplittableDoFn` in ways that can't easily be expressed in terms of a `ParDo`).

tgroh replied:
Moving to `runners-core` also gels with reasoning behind the `GroupByKeyViaGroupAlsoByWindow` expansion being part of `runners-core` - the user has expressed a splittable `DoFn`, and the runner can use the default implementation with minimal wiring by going through an SDK-provided expansion.

jkff replied:
OK, I changed the code to create a ProcessFn at evaluation time. However, I still can't pass StateInternals, TimerInternals and OutputWindowedValue in the constructor due to cyclic dependencies between these and the surrounding objects - e.g. DoFnTester (e.g. the ProcessFn depends on DoFnTester.getTimerInternals(), and DoFnTester obviously depends on the ProcessFn) and ParDoEvaluator (similar cyclic dependency with OutputWindowedValue).

I discussed this at length with @kennknowles and we decided to keep these setters and give up on moving this setup into the ProcessFn's constructor because there's no easy way to do this and because the set-up procedure of a ProcessFn is going to change in nearby PRs anyway (where I'll be adding more runner hooks). But the current PR, I think, is an acceptable intermediate state.

tgroh commented on sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnTester.java:
```diff
@@ -508,6 +517,20 @@ public void onTimer(TimerInternals.TimerData timer) throws Exception {
   }
 
+  @SuppressWarnings({"unchecked", "rawtypes"})
+  public <T> List<WindowedValue<T>> getOrCreateOutput(TupleTag<T> tag) {
```
Unless there's a mirrored `getOuput` which throws if it doesn't exist, this should just be `getOutput`

jkff replied:
Renamed this one and getOutput to respectively getMutableOutput and getImmutableOutput, and fixed a couple more small things in DoFnTester that followed.

jkff commented on runners/direct-java/src/main/java/org/apache/beam/runners/direct/ParDoEvaluatorFactory.java:
```diff
@@ -134,4 +153,8 @@ public void cleanup() throws Exception {
     }
   }
+
+  public DoFnLifecycleManager getFnClone(DoFn<?, ?> fn) {
```
Renamed.
